### ISSUE TITLE
FP S4507 (after the breaking changes in netcore 3.0)

### ISSUE THREAD

**Issue by elukash** [SonarSource/sonar-dotnet#3184]
### Description

After migration to .net core 3.1 Sonar highlighted this snippet with rule:**S4507**

```c#
if (env.IsDevelopment())
           {
                app.UseDeveloperExceptionPage();
            }
```

> Make sure this debug feature is deactivated before delivering the code in production.

### Expected behavior

Code is already deactivated so it should not be highlighted 

### Related information
Seems [this](https://github.com/SonarSource/sonar-dotnet/blob/734df7968037387faade9353c7aedd1a77c010a0/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/Hotspots/DeliveringDebugFeaturesInProduction.cs#L85) matcher is not working correctly now. 

This might be because of in netcore 3.1 the interface IHostingEnvironment has changed to IWebHostEnvironment.

**Comment by andrei-epure-sonarsource** [comment]
@elukash so is it a "false negative" or a "false positive"? You say

> Code is already deactivated so it should not be highlighted

**Comment by andrei-epure-sonarsource** [comment]
Note for us: there is some work to do to enable NetCore 3.1 unit tests. With NetCore 3+ , the `ASP.NET` team changed the packaging and lots of what was available via nuget packages is now available via the SDK.

> Features of ASP.NET Core that were available through one of the packages listed above are available as part of the Microsoft.AspNetCore.App shared framework. The shared framework is the set of assemblies (.dll files) that are installed on the machine and includes a runtime component and a targeting pack. For more information, see The shared framework.

For 3.0, we could reference the [Microsoft.AspNetCore.App.Ref](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Ref/) in our tests, however there's work to be done in the [NugetMetadataFactory](https://github.com/SonarSource/sonar-dotnet/blob/8.4.0.15306/sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/NugetMetadataFactory.cs) to allow this.

The main change is that lots of DLLs, for example `Microsoft.AspNetCore.Http.Extensions.dll` needed by this rule, are present now in `packages\Microsoft.AspNetCore.App.Ref.3.1.0\ref\netcoreapp3.1` (lots of DLLs in this nuget, which is used internally by Microsoft for its SDK) whereas they were present in 2.2.0 in a folder pattern like `packages\Microsoft.AspNetCore.Diagnostics.2.2.0\lib\netstandard2.0` - one DLL per nuget.

This is the line that uses `lib` (valid for 2.2.0), and the UT framework must be extended to search various DLLs in the `Microsoft.AspNetCore.App.Ref.3.1.0\ref` folder.

**Comment by andrei-epure-sonarsource** [comment]
This ticket depends on  #3195

**Comment by elukash** [comment]
> @elukash so is it a "false negative" or a "false positive"? You say
> 
> > Code is already deactivated so it should not be highlighted

This rule shows negative case and this is false, so I call it false negative :)

**Comment by andrei-epure-sonarsource** [comment]
@elukash ok, so it'a actually a False Positive. The terminology we use is the English one ([see Wikipedia](https://en.wikipedia.org/wiki/False_positives_and_false_negatives#False_positive_error)):

> A false positive error, or in short a false positive, commonly called a "false alarm", is a result that indicates a given condition exists, when it does not

**Comment by eric-therond-sonarsource** [comment]
Additional feedback from the community: https://community.sonarsource.com/t/debug-feature-in-production-code-false-positive-s4507/24705

**Comment by pavel-mikula-sonarsource** [comment]
This is duplicate to #3572 that was fixed in 8.13

**Comment by justcoding121** [comment]
If we have an override for `Startup.cs` with `StartupDevelopment.cs`, the overridden file is expected to run only on Development environment. (ASP.Net automatically does that based on ASPNETCORE_ENVIRONMENT="Development"). In production, it won't be called.

![image](https://user-images.githubusercontent.com/2637679/117484554-744fc380-af24-11eb-8fa6-50033cd60ba1.png)


In this case this is a false-positive.
